File: widgets/macos/LyricsAPIService.swift
//
//  LyricsAPIService.swift
//  Daily Lyrics Widget - macOS
//
//  API 통신 로직
//

import Foundation

class LyricsAPIService {
    static let shared = LyricsAPIService()

    // API 베이스 URL (로컬호스트)
    private let baseURL = "http://localhost:58384"

    private init() {}

    /// 현재 가사 가져오기
    func getCurrentLyrics(interval: String = "24h") async -> Result<LyricsData, Error> {
        let urlString = "\(baseURL)/current-lyric?interval=\(interval)"

        guard let url = URL(string: urlString) else {
            return .failure(APIError.invalidURL)
        }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return .failure(APIError.serverError)
            }

            let decoder = JSONDecoder()
            let lyricsResponse = try decoder.decode(LyricsResponse.self, from: data)

            if lyricsResponse.success, let lyricsData = lyricsResponse.data {
                return .success(lyricsData)
            } else {
                return .failure(APIError.noData)
            }
        } catch {
            return .failure(error)
        }
    }

    /// 랜덤 가사 가져오기
    func getRandomLyrics() async -> Result<LyricsData, Error> {
        let urlString = "\(baseURL)/random-lyric"

        guard let url = URL(string: urlString) else {
            return .failure(APIError.invalidURL)
        }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return .failure(APIError.serverError)
            }

            let decoder = JSONDecoder()
            let lyricsResponse = try decoder.decode(LyricsResponse.self, from: data)

            if lyricsResponse.success, let lyricsData = lyricsResponse.data {
                return .success(lyricsData)
            } else {
                return .failure(APIError.noData)
            }
        } catch {
            return .failure(error)
        }
    }

    /// 서버 상태 확인
    func checkHealth() async -> Bool {
        let urlString = "\(baseURL)/health"

        guard let url = URL(string: urlString) else {
            return false
        }

        do {
            let (_, response) = try await URLSession.shared.data(from: url)
            guard let httpResponse = response as? HTTPURLResponse else {
                return false
            }
            return httpResponse.statusCode == 200
        } catch {
            return false
        }
    }
}

/// API 에러 타입
enum APIError: LocalizedError {
    case invalidURL
    case serverError
    case noData
    case networkError

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "잘못된 URL"
        case .serverError:
            return "서버 오류"
        case .noData:
            return "데이터 없음"
        case .networkError:
            return "네트워크 오류"
        }
    }
}

File: widgets/android/LyricsAPIService.kt
package com.example.dailylyrics

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL

/**
 * API 통신 서비스
 */
class LyricsAPIService {
    companion object {
        // API 베이스 URL
        // 실제 기기: PC의 IP 주소로 변경 필요 (예: "http://192.168.0.10:58384")
        // 에뮬레이터: "http://10.0.2.2:58384" 사용 (10.0.2.2는 호스트 PC를 가리킴)
        private const val BASE_URL = "http://10.0.2.2:58384"

        private val json = Json {
            ignoreUnknownKeys = true
            isLenient = true
        }

        /**
         * 현재 가사 가져오기
         */
        suspend fun getCurrentLyrics(interval: String = "24h"): Result<LyricsData> {
            return withContext(Dispatchers.IO) {
                try {
                    val url = URL("$BASE_URL/current-lyric?interval=$interval")
                    val connection = url.openConnection() as HttpURLConnection
                    connection.requestMethod = "GET"
                    connection.connectTimeout = 5000
                    connection.readTimeout = 5000

                    if (connection.responseCode == HttpURLConnection.HTTP_OK) {
                        val response = connection.inputStream.bufferedReader().use { it.readText() }
                        val lyricsResponse = json.decodeFromString<LyricsResponse>(response)

                        if (lyricsResponse.success && lyricsResponse.data != null) {
                            Result.success(lyricsResponse.data)
                        } else {
                            Result.failure(Exception(lyricsResponse.error ?: "No data"))
                        }
                    } else {
                        Result.failure(Exception("Server error: ${connection.responseCode}"))
                    }
                } catch (e: Exception) {
                    Result.failure(e)
                }
            }
        }

        /**
         * 랜덤 가사 가져오기
         */
        suspend fun getRandomLyrics(): Result<LyricsData> {
            return withContext(Dispatchers.IO) {
                try {
                    val url = URL("$BASE_URL/random-lyric")
                    val connection = url.openConnection() as HttpURLConnection
                    connection.requestMethod = "GET"
                    connection.connectTimeout = 5000
                    connection.readTimeout = 5000

                    if (connection.responseCode == HttpURLConnection.HTTP_OK) {
                        val response = connection.inputStream.bufferedReader().use { it.readText() }
                        val lyricsResponse = json.decodeFromString<LyricsResponse>(response)

                        if (lyricsResponse.success && lyricsResponse.data != null) {
                            Result.success(lyricsResponse.data)
                        } else {
                            Result.failure(Exception(lyricsResponse.error ?: "No data"))
                        }
                    } else {
                        Result.failure(Exception("Server error: ${connection.responseCode}"))
                    }
                } catch (e: Exception) {
                    Result.failure(e)
                }
            }
        }

        /**
         * 서버 상태 확인
         */
        suspend fun checkHealth(): Boolean {
            return withContext(Dispatchers.IO) {
                try {
                    val url = URL("$BASE_URL/health")
                    val connection = url.openConnection() as HttpURLConnection
                    connection.requestMethod = "GET"
                    connection.connectTimeout = 3000
                    connection.readTimeout = 3000

                    connection.responseCode == HttpURLConnection.HTTP_OK
                } catch (e: Exception) {
                    false
                }
            }
        }
    }
}
